hw/drivers/flash/ipc_nrf5340_flash/src/ipc_nrf5340_flash.c

```diff
@@ -0,0 +1,580 @@
+    server_resp.hdr.msg_len = server_resp.header_len + server_resp.data_len;
+
+    /* Send part included in ipc_msg structure. */
+    ipc_nrf5340_send(SERVER_OUT_CHANNEL, &server_resp, server_resp.header_len);
+    if (server_resp.data_len > 0) {
+        /* Send part passed by pointer if any. */
+        ipc_nrf5340_send(SERVER_OUT_CHANNEL, server_resp.data, server_resp.data_len);
+    }
+
+    req->hdr.type = 0;
+
+    return 0;
+}
+
+static int
+nrf5340_ipc_flash_std_resp(struct ipc_msg *req, int rc, void *data, uint32_t data_len)
+{
+    server_resp.header_len = HDR_SZ;
+    server_resp.data_len = data_len;
+    server_resp.data = data;
+
+    /* Add rc to header, this updates header_len */
+    put32(&server_resp, (uint32_t)rc);
+
+    return nrf5340_ipc_flash_resp(req);
+}
+
+static int
+nrf5340_ipc_flash_info_resp(struct ipc_msg *req)
+{
+    return nrf5340_ipc_flash_std_resp(req, 0, (void *)&nrf5340_flash_dev, sizeof(nrf5340_flash_dev));
+}
+
+static int
+nrf5340_ipc_flash_read_resp(struct ipc_msg *req)
+{
+    const uint8_t *p;
+    uint32_t address;
+    uint32_t size;
+
+    assert(req->hdr.msg_len = HDR_SZ + 2 * sizeof(uint32_t));
+    nrf5340_ipc_flash_server_read(&req->param, req->hdr.msg_len - HDR_SZ);
+
+    p = req->param;
+    address = get32(&p);
+    size = get32(&p);
+
+    return nrf5340_ipc_flash_std_resp(req, 0, (void *)address, size);
+}
+
+static void
+flash_write_cb(struct os_event *event)
+{
+    int rc = 0;
+    struct ipc_msg *req = event->ev_arg;
+    uint32_t left;
+    uint32_t address;
+    uint32_t chunk;
+
+    nrf5340_ipc_flash_server_read(&address, sizeof(uint32_t));
+    nrf5340_ipc_flash_server_read(&left, sizeof(uint32_t));
+    assert(left == req->hdr.msg_len - (HDR_SZ + 2 * sizeof(uint32_t)));
+
+    while (left && rc == 0) {
+        chunk = sizeof(req->param);
+        if (chunk > left) {
+            chunk = left;
+        }
+        nrf5340_ipc_flash_server_read(req->param, chunk);
+        rc = hal_flash_write(0, address, req->param, chunk);
+        left -= chunk;
+        address += chunk;
+    }
+
+    nrf5340_ipc_flash_std_resp(req, rc, NULL, 0);
+}
+
+static struct os_event flash_write_event = {
+    .ev_cb = flash_write_cb,
+};
+
+static void
+nrf5340_ipc_flash_server_write(struct ipc_msg *req)
+{
+    flash_write_event.ev_arg = req;
+    os_eventq_put(os_eventq_dflt_get(), &flash_write_event);
+}
+
+static void
+flash_erase_cb(struct os_event *event)
+{
+    int rc;
+    struct ipc_msg *req = event->ev_arg;
+    const uint8_t *p = req->param;
+
+    rc = hal_flash_erase_sector(0, get32(&p));
+
+    nrf5340_ipc_flash_std_resp(req, rc, NULL, 0);
+}
+
+static struct os_event erase_event = {
+    .ev_cb = flash_erase_cb,
+};
+
+static void
+nrf5340_ipc_flash_server_erase(struct ipc_msg *req)
+{
+    nrf5340_ipc_flash_server_read(&req->param, sizeof(uint32_t));
+    erase_event.ev_arg = req;
+    os_eventq_put(os_eventq_dflt_get(), &erase_event);
```

Comment:
So, I think it would be better to make this eventq configurable. We do not want this driver to work on the main eventq always.